oximeter/api/src/lib.rs

```diff
@@ -64,3 +75,120 @@ pub struct CollectorInfo {
+    pub address: SocketAddr,
+
+    /// The time the producer was first registered with us.
+    pub registered: DateTime<Utc>,
+
+    /// The last time the producer's information was updated.
+    pub updated: DateTime<Utc>,
+
+    /// The last time we started to collect from this producer.
+    ///
+    /// This is None if we've never attempted to collect from the producer.
+    pub last_collection_started: Option<DateTime<Utc>>,
+
+    /// The last time we successfully completed a collection from this producer.
+    ///
+    /// This is None if we've never successfully collected from the producer.
+    /// Note that this can be before `last_collection_started`, when we're in
+    /// the middle of a collection or the last collection has failed.
+    pub last_successful_collection: Option<DateTime<Utc>>,
+
+    /// The number of samples collected in the last successful collection.
+    pub n_samples_in_last_collection: Option<u64>,
+
+    /// The duration of the last successful collection.
+    pub last_collection_duration: Option<Duration>,
+
+    /// The last time we failed to collect from this producer.
+    ///
+    /// This is None if we've never failed to collect from the producer. Note
+    /// that this can be before `last_collection_started`, when we're in the
+    /// middle of a collection or the last collection was successful.
+    pub last_failed_collection: Option<DateTime<Utc>>,
+
+    /// A string describing why the last collection failed.
+    pub last_failure_reason: Option<String>,
+
+    /// The total number of successful collections we've made.
+    pub n_collections: u64,
+
+    /// The total number of failed collections.
+    pub n_failures: u64,
+}
+
+impl ProducerDetails {
+    pub fn new(info: &ProducerEndpoint) -> Self {
+        let now = Utc::now();
+        Self {
+            id: info.id,
+            interval: info.interval,
+            address: info.address,
+            registered: now,
+            updated: now,
+            last_collection_started: None,
+            last_successful_collection: None,
+            n_samples_in_last_collection: None,
+            last_collection_duration: None,
+            last_failed_collection: None,
+            last_failure_reason: None,
+            n_collections: 0,
+            n_failures: 0,
+        }
+    }
+
+    /// Update with new producer information.
+    ///
+    /// # Panics
+    ///
+    /// This panics if the new information refers to a different ID.
+    pub fn update(&mut self, new: &ProducerEndpoint) {
+        assert_eq!(self.id, new.id);
+        self.updated = Utc::now();
+        self.address = new.address;
+        self.interval = new.interval;
+    }
+
+    /// Update when we start a collection
+    pub fn start_collection(&mut self) {
+        self.last_collection_started = Some(Utc::now());
+    }
+
+    /// Update when we successfully complete a collection.
+    ///
+    /// # Panics
+    ///
+    /// This panics if no collection was started.
```

Comment:
Hm, this seems okay but not ideal. Could `start_collection()` return some kind of handle that could turn into success/fail and statically avoid this panic? Something like

```rust
let handle = details.start_collection();
match do_collection() {
    Ok(_) => details.end_collection(handle.success(n_samples)),
    Err(_) => details.end_collection(handle.failure("failure reason")),
}
```

If the handle held the start time internally, that would also avoid incorrect timings caused by incorrectly paired calls; e.g.,

1. `start_collection()`
2. `start_collection()`
3. `on_success()` 
4. `on_success()`

This is a little questionable given we serialize collection requests today, but I think if a caller did this both success calls would calculate time based on when the second collection started, which is probably not what they intended.